File: compiler/src/test/java/com/aaroncoplan/waterfall/compiler/tests/EndToEndSmokeTest.java
package com.aaroncoplan.waterfall.compiler.tests;

import com.aaroncoplan.waterfall.compiler.Main;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import java.io.ByteArrayOutputStream;
import java.io.PrintStream;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

/**
 * Smoke tests that exercise the compiler end-to-end on the example files.
 * If something in the AST -> codegen path regresses badly, these break first.
 */
public class EndToEndSmokeTest {

    private PrintStream originalOut;
    private PrintStream originalErr;
    private ByteArrayOutputStream capturedOut;

    @Before
    public void redirectStdout() {
        originalOut = System.out;
        originalErr = System.err;
        capturedOut = new ByteArrayOutputStream();
        System.setOut(new PrintStream(capturedOut));
        // Suppress logger noise on stderr to keep the test report clean.
        System.setErr(new PrintStream(new ByteArrayOutputStream()));
    }

    @After
    public void restoreStdout() {
        System.setOut(originalOut);
        System.setErr(originalErr);
    }

    @Test
    public void functionWithBodyTranslates() {
        Main.main(new String[]{"../examples/FunctionWithBodyModule.wf"});
        String output = capturedOut.toString();
        assertFalse("Output should be non-empty", output.isEmpty());
        assertTrue("Output should declare flag variable", output.contains("int flag = 0;"));
        assertTrue("Output should declare doSomething function", output.contains("void doSomething()"));
        assertTrue("Output should contain untyped declaration inferred as int", output.contains("int x = 14;"));
    }

    @Test
    public void controlFlowTranslates() {
        Main.main(new String[]{"../examples/ControlFlowModule.wf"});
        String output = capturedOut.toString();
        assertTrue("Output should emit if-block", output.contains("if ("));
        assertTrue("Output should emit else-if-block", output.contains("else if ("));
        assertTrue("Output should emit else-block", output.contains("else {"));
        assertTrue("Output should emit for-block", output.contains("for ("));
        assertTrue("Output should emit local function call",
                output.contains("doSomething()"));
        assertTrue("Output should emit module-qualified function call as Module_fn",
                output.contains("Other_helper(1, 2)"));
    }

    @Test
    public void variableDeclarationsTranslates() {
        Main.main(new String[]{"../examples/VariableDeclarationsModule.wf"});
        String output = capturedOut.toString();
        assertTrue("Output should declare int x = 4;", output.contains("int x = 4;"));
    }

    @Test
    public void emptyModuleProducesNoCode() {
        Main.main(new String[]{"../examples/EmptyModule.wf"});
        String output = capturedOut.toString().trim();
        assertTrue("Empty module should produce empty output, got: [" + output + "]", output.isEmpty());
    }
}

File: compiler/src/test/kotlin/com/aaroncoplan/waterfall/compiler/tests/EndToEndSmokeTest.kt
package com.aaroncoplan.waterfall.compiler.tests

import com.aaroncoplan.waterfall.compiler.CompilerError
import com.aaroncoplan.waterfall.compiler.Main
import org.junit.After
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import java.io.ByteArrayOutputStream
import java.io.PrintStream

/**
 * Smoke tests that exercise the compiler end-to-end on the example files.
 * If something in the AST -> codegen path regresses badly, these break first.
 */
class EndToEndSmokeTest {

    private lateinit var originalOut: PrintStream
    private lateinit var originalErr: PrintStream
    private lateinit var capturedOut: ByteArrayOutputStream

    @Before
    fun redirectStdout() {
        originalOut = System.out
        originalErr = System.err
        capturedOut = ByteArrayOutputStream()
        System.setOut(PrintStream(capturedOut))
        // Suppress logger noise on stderr to keep the test report clean.
        System.setErr(PrintStream(ByteArrayOutputStream()))
    }

    @After
    fun restoreStdout() {
        System.setOut(originalOut)
        System.setErr(originalErr)
    }

    @Test
    fun functionWithBodyTranslates() {
        Main.run(arrayOf("--target", "js", "../examples/FunctionWithBodyModule.wf"))
        val output = capturedOut.toString()
        assertFalse("Output should be non-empty", output.isEmpty())
        assertTrue("Output should declare flag variable", output.contains("let flag = 0;"))
        assertTrue("Output should declare doSomething function", output.contains("function doSomething()"))
        assertTrue("Output should contain untyped declaration inferred as int", output.contains("let x = 14;"))
    }

    @Test
    fun controlFlowModuleFailsVerificationInP11() {
        // P11 §4.1 (OQ-11.6=strict): ControlFlowModule.wf references undeclared identifiers:
        // `things` in for-loop collection and `doSomething` as a LOCAL function call.
        // The verifier now rejects this module and throws CompilerError.
        // Golden files for ControlFlowModule are updated to empty string per §7.2.
        try {
            Main.run(arrayOf("--target", "js", "../examples/ControlFlowModule.wf"))
            org.junit.Assert.fail(
                "Expected CompilerError: ControlFlowModule.wf should fail to compile in P11 " +
                "(undeclared 'things' in for-collection + undeclared 'doSomething' in LOCAL call)"
            )
        } catch (e: CompilerError) {
            // Expected — verifier rejects undeclared identifiers per OQ-11.6=strict
        }
    }

    @Test
    fun variableDeclarationsTranslates() {
        Main.run(arrayOf("--target", "js", "../examples/VariableDeclarationsModule.wf"))
        val output = capturedOut.toString()
        assertTrue("Output should declare let x = 4;", output.contains("let x = 4;"))
    }

    @Test
    fun emptyModuleHasNoDeclarations() {
        Main.run(arrayOf("--target", "js", "../examples/EmptyModule.wf"))
        val output = capturedOut.toString().trim()
        // JS backend emits a module comment header; assert output contains only the header line.
        assertTrue("Empty module output should contain module comment, got: [$output]",
            output.contains("// module EmptyModule"))
        assertFalse("Empty module output should contain no function or variable declarations",
            output.contains("function ") || output.contains("let ") || output.contains("const "))
    }

    @Test
    fun defaultTargetIsJs() {
        Main.run(arrayOf("../examples/FibonacciModule.wf"))   // no --target
        val output = capturedOut.toString()
        assertTrue("Default target should produce JS output",
            output.contains("function fib(") || output.contains("let total = 0;"))
    }
}
